src/controllers/authenticationController.ts

```diff
@@ -105,6 +108,80 @@ export default class AuthenticationController {
       .send(metadata);
   }
 
+  /**
+   * Returns data about the current user session.
+   */
+  public async getSessionState(
+    req: express.Request
+  ): Promise<Either<Error, ISessionState>> {
+    try {
+      const maybeToken = await this.extractTokenFromRequest(req);
+      if (isNone(maybeToken)) {
+        return left(new Error("No token in the request"));
+      }
+
+      const errorOrSession = await this.sessionStorage.get(maybeToken.value);
+      if (isLeft(errorOrSession)) {
+        const errorOrSessionState = await this.refreshUserToken(
+          maybeToken.value
+        );
+
+        if (isLeft(errorOrSessionState)) {
+          const error = errorOrSessionState.value;
+          return left(error);
+        }
+
+        const sessionState = errorOrSessionState.value;
+        return right(sessionState);
+      }
+
+      const session = errorOrSession.value;
+      const publicSession = {
+        expireAt: session.expireAt,
+        expired: session.expired
+      };
+      return right(publicSession);
+    } catch (err) {
+      return left(new Error("Error while loading the session"));
+    }
+  }
+
+  /**
+   * Extracts the Bearer token from the authorization header in the request.
+   */
+  private async extractTokenFromRequest(
+    req: express.Request
+  ): Promise<Option<string>> {
+    if (req.headers && req.headers.authorization) {
```

Comment:
if you add `&& typeof req.headers.authorization === "string"` you will be sure this is a string and you can skip the cast in the next line (the TS type system will know you have checked already that is a string), see [type guards](https://basarat.gitbooks.io/typescript/docs/types/typeGuard.html)

Reply:
I've already tried that way but TS warns for authorization: `Object is possibly 'undefined'` and for split: `Property 'split' does not exist on type 'string | string[]'`

Reply:
that error should warn you that that value could be either a string or an array of strings, you instead are forcing a cast to a string - so in case it's an array, the `split` will throw an exception

Reply:
you shouldn't get an `Object is possibly 'undefined'` if you have all three checks:

```ts
if (req.headers && req.headers.authorization && typeof req.headers.authorization === "string")
```

Reply:
![screen shot 2018-03-23 at 13 01 50](https://user-images.githubusercontent.com/436670/37828140-647f1b1c-2e9a-11e8-8d6c-0a0f499e4125.png)

Reply:
this works:

```ts
interface IReq {
  headers?: {
    authorization?: string | string[]
  }
};

function f(r: IReq): void {
  if (r.headers && r.headers.authorization && typeof r.headers.authorization === "string") {
    const h: string = r.headers.authorization;
  }
}
```